### PR TITLE
[spirv] Non-floating point matrix layout.

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

**Comment by ehsannas** [comment]
Looks like the buildbot hit the 1hour limit again.

**Comment by ehsannas** [comment]
Rebased.

File: tools/clang/test/CodeGenSPIRV/vk.layout.sbuffer.nested.std430.hlsl [hunk 1]
```diff
@@ -6,36 +6,40 @@ struct R {                         // Alignment    Offset  Size
...
+    row_major    int3x2   sf7[2];  // 8(vec2)  -> 1744  + 2(array) * stride(3 * 8(vec2))  = 1792
+                 float    sf8;     // 4        -> 1792  + 4                               = 1796
+};                                 // 16(max)                                               1808 (1796 round up to S alignment)
 
 struct T {        // Alignment    Offset  Size              Next
-    R    tf1[2];  // 16        -> 0     + 2(array) * 256  = 512
-    S    tf2[3];  // 16        -> 512   + 3(array) * 1376 = 4640
-    uint tf3;     // 4         -> 4640  + 4               = 4644
-};                // 16(max)                                4656 (4640 round up to T alignment)
+    R    tf1[2];  // 16        -> 0     + 2(array) * 352  = 704
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/vk.layout.sbuffer.nested.std430.hlsl]: Awesome! :)

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 2]
```diff
@@ -746,6 +748,36 @@ bool TypeTranslator::isMxNMatrix(QualType type, QualType *elemType,
...
+
+  if (const auto *arrayType = astContext.getAsConstantArrayType(type)) {
+    if (isMxNMatrix(arrayType->getElementType(), &elemType) &&
+        !elemType->isFloatingType())
+      return isColMajorDecl(decl);
+  }
+
+  if (const auto *structType = type->getAs<RecordType>()) {
+    const auto *decl = structType->getDecl();
+    // Create fields for all members of this struct
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Oops!

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 3]
```diff
@@ -746,6 +748,36 @@ bool TypeTranslator::isMxNMatrix(QualType type, QualType *elemType,
   return true;
 }
 
+bool TypeTranslator::isNonFpColMajorMatrix(QualType type,
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Done

**Pull request merged**